### PR TITLE
Add pattern and executor for ConvBiasResAddActiv GraphAPI

### PR DESCRIPTION
**Pull request by BrianHarrisonAMD**
This change adds the pattern and executor for ConvBiasResAddActiv GraphAPI, and enables previously disabled tests to verify functionality.

Changes:
- Added Executor, pattern, and hooked them up inside find_engine
- Updated test cases to run pattern matching, executor, and verify output
- Tested with existing test cases
- Updated GraphApiVariantPack tests to allow nullptr for workspace

### CODE REVIEW

File: src/graphapi/find_engine.cpp
Comment (BrianHarrisonAMD) [inline src/graphapi/find_engine.cpp]: Could be swapped to MIOPEN_THROW_IF

File: src/graphapi/conv_bias_res_add_activ_forward_executor.cpp
Comment (BrianHarrisonAMD) [inline src/graphapi/conv_bias_res_add_activ_forward_executor.cpp]: Looking at this again, I think it would make more sense to move some of this into the GraphPatternMatcher, and instead of passing the operations, I can pass the values I need out of them.  

For example, instead of passing the addOp, and needing to find which tensor to use for z, I could do that in the GraphPatternMatcher, and then make the executor's constructor explicit. I think that fits the design better, and fits closer to the other executor.

File: src/graphapi/conv_bias_res_add_activ_forward_executor.cpp
Comment (BrianHarrisonAMD) [inline src/graphapi/conv_bias_res_add_activ_forward_executor.cpp]: I think we should probably be checking for this in the pattern matcher, and throw if they aren't evenly divisible.

File: src/graphapi/find_engine.cpp
Comment (BrianHarrisonAMD) [inline src/graphapi/find_engine.cpp]: Come to think of it, should this check actually be part of the matches check above?

I think that might actually make more sense, since we don't actually have a bias node for pattern checking, and this pattern requires a bias node that needs to be checked based off the provided tensor descriptions.

For example, this pattern matcher could currently match to a graph with the same nodes, but want to perform 2 adds instead of a bias. It would then throw when asking for the engines since there is no bias node, and that seems incorrect.

(Looking for feedback from reviewers on this)

File: src/graphapi/find_engine.cpp
Comment (BrianHarrisonAMD) [inline src/graphapi/find_engine.cpp:163]: Should these checks be part of the matches check?
Not sure if matches implies that it should run without issue, or just if the pattern matches. 

In this case, I think it makes sense not to be part of the matches check since this implies invalid tensors were passed in for the ConvolutionOp. 

(Looking for feedback from reviewers on this)

File: src/graphapi/find_engine.cpp
Comment (BrianHarrisonAMD) [inline src/graphapi/find_engine.cpp]: Also, need to add tests that check for these cases once it's determined how we want to handle it.

Should make sure it's throwing as expected, or not matching as expected once we know how to handle these cases.

File: src/graphapi/conv_bias_res_add_activ_forward_executor.cpp
Comment (amberhassaan) [inline src/graphapi/conv_bias_res_add_activ_forward_executor.cpp]: we should use a `static_cast` here.

Comment (amberhassaan) [reply src/graphapi/conv_bias_res_add_activ_forward_executor.cpp]: May be put an assert that value is less than max int.

File: src/graphapi/find_engine.cpp
Comment (amberhassaan) [reply src/graphapi/find_engine.cpp:163]: some of these checks are in the high-level convolution API but then fused convolutions don't seem to go through that API and hence bypass those checks. I don't think we need to check extensively here and the solver should do more checks of the nature of valid params. Here we primarily check the graph (nodes and tensors). This is my initial opinion subject to change.

File: src/graphapi/find_engine.cpp
Comment (amberhassaan) [inline src/graphapi/find_engine.cpp]: I don't know if there's flexibility about the position of Bias node. Usually Bias follows Conv directly.

File: src/graphapi/find_engine.cpp
Comment (amberhassaan) [inline src/graphapi/find_engine.cpp]: What's this logic doing? When would `X` be virtual?

**Review by amberhassaan — APPROVED** [review]
Some comments but looks OK overall.

File: src/graphapi/find_engine.cpp
Comment (BrianHarrisonAMD) [reply src/graphapi/find_engine.cpp]: I tried to make it flexible as I wasn't sure if we wanted to be too strict with which input is the virtual one.

For example, if the graph specified that B is the virtual input instead of X, then I think it should be able to handle that still since order doesn't matter for add.

File: src/graphapi/find_engine.cpp
Comment (BrianHarrisonAMD) [reply src/graphapi/find_engine.cpp]: I was wondering about that.
I figured since they were the same node type, and that they are both adds, then it would be more flexible this way, but I could swap it to enforce bias after conv if that's better.

File: src/graphapi/find_engine.cpp
Comment (BrianHarrisonAMD) [reply src/graphapi/find_engine.cpp:163]: In this specific case, since I need to calculate groupCount for the convolution descriptor , I think I need to check and throw here before calculating the groupCount, but I like the general opinion of how we should handle verification. That will minimize duplicating existing validity checks, and seems like a good approach to me.

File: src/graphapi/find_engine.cpp
Comment (BrianHarrisonAMD) [reply src/graphapi/find_engine.cpp]: Decided to move it since I think it's better not to match if the bias node isn't there.

File: src/include/miopen/graphapi/variant_pack.hpp
Comment (randyspauldingamd) [inline src/include/miopen/graphapi/variant_pack.hpp:161]: Just to verify intention--now the user _must_ call this function, but it's ok if they set it to NULL.

File: src/graphapi/find_engine.cpp
Comment (randyspauldingamd) [inline src/graphapi/find_engine.cpp]: This swaps alpha1 and alpha2 when X is concrete (i.e. B is virtual). Just a wild guess, but it seems counterintuitive. Also, I find this code to be (subjectively ofc) easier to read:
```suggestion
            auto alpha1 = std::visit(convertToFloat, add->getAlpha1();
            auto alpha2 = std::visit(convertToFloat, add->getAlpha2();
            if(addXIsVirtual) std::swap(alpha1, alpha2);
```

File: src/graphapi/find_engine.cpp
Comment (randyspauldingamd) [inline src/graphapi/find_engine.cpp:186]: It's sure tempting to add a wrapper class that encapsulates all this selection. Something like:
```suggestion
class OperationPointwise_OneVirtual   // TODO: better name
{ // forgive the pseudocode and partial code--hopefully it'll do for a quick suggestion
public:
  Tensor *add, *bias;
  float alpha1, alpha2;
  OperationPointwise_OneVirtual(OpGraph* graph_ptr) 
{
  // do sanity checking along the way
  add = // identify the 'add' node, identify its concrete Tensor, and store
  bias = // identify the 'bias' node, identify its concrete Tensor, and store
  // select and store alpha1 and alpha2
}
  float getAlpha1() { return alpha1; }
  float getAlpha2() { return alpha2; }
};
```
Then this function becomes, starting at 146: 
```suggestion
OperationPointwise_OneVirtual oneVirtual{graph_ptr};
std::shared_ptr<GraphPatternExecutor> exec = ConvBiasResAddActivForwardExecutor::make(
            conv->getX(),
            conv->getW(),
            conv->getConvolution(),
            groupCount,
            oneVirtual.add,
            oneVirtual.bias,
            activ->getY(),
            oneVirtual.getAlpha1(),
            oneVirtual.getAalpha2(),
            std::visit(convertToFloat, activ->getAlpha1()));
```

File: src/CMakeLists.txt
Comment (randyspauldingamd) [inline src/CMakeLists.txt]: Please keep filenames sorted alphabetically. Although I don't know the convention we use for underscore, since it's ascii code falls between the capitals and lowercase letters :)

File: src/graphapi/find_engine.cpp
Comment (amberhassaan) [reply src/graphapi/find_engine.cpp]: ok, I see. Good to keep the flexibility.

File: src/CMakeLists.txt
Comment (amberhassaan) [reply src/CMakeLists.txt]: agree. Sorting should be maintained.

File: src/graphapi/find_engine.cpp
Comment (CAHEK7) [reply src/graphapi/find_engine.cpp:163]: I guess one you've found a list of applicable engines, you have to go through that list and check `isApplicable` method.
I'm a bit out of the context and probably it's not possible at this step, but we can't return non-applicable solver\engine in any case.
(I "unresolved" this case just to make it visible)

File: src/CMakeLists.txt
Comment (CAHEK7) [reply src/CMakeLists.txt]: Does anybody know any linter for CMAKE?

File: src/graphapi/conv_bias_res_add_activ_forward_executor.cpp
Comment (CAHEK7) [inline src/graphapi/conv_bias_res_add_activ_forward_executor.cpp]: I would recommend keeping these functions inside an unnamed namespace, since it covers more cases that just a `static`. It's my personal preference and recommendation, but since we don't have any particular style guide, we can do whatever we want.

File: src/graphapi/find_engine.cpp
Comment (CAHEK7) [inline src/graphapi/find_engine.cpp]: It looks like https://en.cppreference.com/w/cpp/algorithm/count

File: src/graphapi/find_engine.cpp
Comment (CAHEK7) [inline src/graphapi/find_engine.cpp]: We can do `dynamic_cast` to a reference and it will automatically throw an exception.
But probably we've always known that all the `dynamic_cast`'s are castable and they have never returned `nullptr`.

File: src/graphapi/find_engine.cpp
Comment (CAHEK7) [inline src/graphapi/find_engine.cpp]: Some sort of nitpicking:
```suggestion
    std::string_view name() const final
    {
        static const std::string_view n{"convbiasresaddactivation_fwd"};
        return n;
    }
```
It's identical even on Og, but slightly better on O0.
And it expresses the intension clearly - you don't need to "construct" `string_view` from the `const char*`

File: src/graphapi/find_engine.cpp
Comment (CAHEK7) [inline src/graphapi/find_engine.cpp:141]: Is it an assertion of a fatal error which should be properly reported even in release build?

File: src/include/miopen/graphapi/conv_bias_res_add_activ_forward_executor.hpp
Comment (CAHEK7) [inline src/include/miopen/graphapi/conv_bias_res_add_activ_forward_executor.hpp:80]: Can it depend on a particular matched engine?

File: test/gtest/graphapi_conv_bias_res_add_activ_fwd.cpp
Comment (CAHEK7) [inline test/gtest/graphapi_conv_bias_res_add_activ_fwd.cpp:469]: Can this test be a part of the common testsuite? Like
```C++
if (input_c % wei_c != 0)
    ASSERT_THROW(engines = gr::findEngines(&graph), miopen::Exception);
    SUCCEED();
    return;
else
    ASSERT_NO_THROW(engines = gr::findEngines(&graph), miopen::Exception);

...the rest of the test...
```
It eliminates extra initialization code.

**Review by CAHEK7 — COMMENTED** [review]
There are a few questions and recommendations.
I don't block this PR, it can be fixed later, or right now, or even can be ignored.

**Review by junliume — APPROVED** [review]

**Comment by junliume** [comment]
@CAHEK7 last round call for reviews, if needed @BrianHarrisonAMD please deal with the review opinions in a follow up PR. Thanks!

File: src/include/miopen/graphapi/variant_pack.hpp
Comment (BrianHarrisonAMD) [reply src/include/miopen/graphapi/variant_pack.hpp:161]: Since workspace was defaulting to nullptr before, I believe they always had to set it, but now it allows them to set a null value.

File: src/CMakeLists.txt
Comment (BrianHarrisonAMD) [reply src/CMakeLists.txt]: Nice catch!

Will do.

File: src/graphapi/find_engine.cpp
Comment (BrianHarrisonAMD) [reply src/graphapi/find_engine.cpp]: Ah good point.
I can remove this.

File: src/graphapi/find_engine.cpp
Comment (BrianHarrisonAMD) [reply src/graphapi/find_engine.cpp:141]: I think in this case it will throw if it doesn't match, (since the code below will fail to find, and explode), and we don't want to pay the runtime cost of checking matches again. I was following how the other pattern matcher was set up, and I think that is the reason anyways.

File: src/graphapi/find_engine.cpp
Comment (BrianHarrisonAMD) [reply src/graphapi/find_engine.cpp:163]: I think in this case, it is "applicable", but these exceptions would indicate improperly set up convolution operations. That's why I decided not to move these into the matches check anyways. Basically, if this were to throw, then I would expect any pattern matcher / executor would eventually throw since it's a malformed request.

File: src/graphapi/find_engine.cpp
Comment (BrianHarrisonAMD) [reply src/graphapi/find_engine.cpp:186]: Yea that's a good point. There is a lot of virtual checks sprinkled around the pattern checker code, and adding a quick utility class to manage that would be cleaner. 

Thanks!

File: src/include/miopen/graphapi/conv_bias_res_add_activ_forward_executor.hpp
Comment (BrianHarrisonAMD) [reply src/include/miopen/graphapi/conv_bias_res_add_activ_forward_executor.hpp:80]: Eventually probably, but with this current implementation, it will always call the fusion method which doesn't (currently) use workspace.

File: test/gtest/graphapi_conv_bias_res_add_activ_fwd.cpp
Comment (BrianHarrisonAMD) [reply test/gtest/graphapi_conv_bias_res_add_activ_fwd.cpp:469]: That's a good point, ill try to combine these.

Thanks!

Comment (BrianHarrisonAMD) [reply test/gtest/graphapi_conv_bias_res_add_activ_fwd.cpp:469]: Oh right I remember now, the reason I can't combine these is that it will throw earlier if I try to do the same flow with improperly formed arguments.

I think If I were to try and combine them, then the initialization would diverge quite a bit, and think it's better to keep it like this.

**Comment by BrianHarrisonAMD** [comment]
Code review comments should be addressed now, and just waiting for CI pass. 

Thanks for all the feedback!

**Review by CAHEK7 — APPROVED** [review]

**Comment by junliume** [comment]
Updated by resolving conflicts, will merge when CI passes again. Thanks!

**Review by junliume — APPROVED** [review]

**Pull request merged**
